src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile_isolated_strategy.py

```diff
@@ -5,33 +5,201 @@
+
+    def all_successful(self, keys):
+      return all(stat == SUCCESS for stat in [self._statuses[k] for k in keys])
+
+  def __init__(self, parent_work_unit, run_tracker, worker_count, log):
+    self._log = log
+    self._parent_work_unit = parent_work_unit
+    self._run_tracker = run_tracker
+    self._worker_count = worker_count
+    self._dependees = defaultdict(list)
+    self._work = {}
+    self._work_keys_as_scheduled = []
+
+  def log_dot_graph(self):
+    for key in self._work_keys_as_scheduled:
+      self._log.debug("{} -> {{\n  {}\n}}".format(key, ',\n  '.join(self._dependees[key])))
+
+  def schedule(self, key, fn, dependency_keys, on_success=None, on_failure=None):
+    """Inserts work into the execution graph with its dependencies.
+
+    Assumes dependencies have already been scheduled, and raises an error otherwise."""
+    self._work_keys_as_scheduled.append(key)
+    self._work[key] = self.Work(fn, dependency_keys, on_success, on_failure)
+    for dep_name in dependency_keys:
+      if dep_name not in self._work:
+        raise Exception("Expected {} not scheduled before dependent {}".format(dep_name, key))
+      self._dependees[dep_name].append(key)
+
+  def find_work_without_dependencies(self):
+    # Topo sort doesn't mean all no-dependency targets are listed first,
+    # so we look for all work without dependencies
+    return filter(
+      lambda key: len(self._work[key].dependencies) == 0, self._work_keys_as_scheduled)
+
+  def execute(self):
+    """Runs scheduled work, ensuring all dependencies for each element are done before execution.
+
+    spawns a work pool of the specified size.
+    submits all the work without any dependencies
+    when a unit of work finishes,
+      if it is successful
+        calls success callback
+        checks for dependees whose dependencies are all successful, and submits them
+      if it fails
+        calls failure callback
+        marks dependees as failed and queues them directly into the finished work queue
+    when all work is either successful or failed,
+      cleans up the work pool
+    if there's an exception on the main thread,
+      calls failure callback for unfinished work
+      aborts work pool
+      re-raises
+    """
+    self.log_dot_graph()
+
+    status_table = self.StatusTable(self._work_keys_as_scheduled)
+    finished_queue = queue.Queue()
+
+    work_without_dependencies = self.find_work_without_dependencies()
+    if len(work_without_dependencies) == 0:
+      raise self.ExecutionFailure("No work without dependencies! There must be a "
+                                  "circular dependency")
+
+    def worker(work_key, work):
+      try:
+        work()
+        result = (work_key, True, None)
+      except Exception as e:
+        result = (work_key, False, e)
+      finished_queue.put(result)
+
+    pool = WorkerPool(self._parent_work_unit, self._run_tracker, self._worker_count)
```

Comment:
No, because I want control over the number of workers, and also because I need to be sure that all my work is done before closing out the chunk. I played with using the global background worker pool earlier and ran into some awkwardness around that.